packages/pwa/app/commerce-api/contexts.js

```diff
@@ -0,0 +1,163 @@
+ * }
+ *
+ */
+
+/************ Basket ************/
+export const BasketContext = createContext()
+export const BasketProvider = BasketContext.Provider
+
+/************ Customer ************/
+export const CustomerContext = createContext()
+export const CustomerProvider = CustomerContext.Provider
+
+/************ Customer Product Lists ************/
+const CPLInitialValue = {
+    productLists: {
+        // this is a map of product lists
+        // keyed by list id
+    }
+}
+const CPLActionTypes = {
+    RECEIVE_LISTS: 'RECEIVE_LISTS',
+    RECEIVE_LIST: 'RECEIVE_LIST',
+    CREATE_LIST_ITEM: 'CREATE_LIST_ITEM',
+    UPDATE_LIST_ITEM: 'UPDATE_LIST_ITEM',
+    REMOVE_LIST_ITEM: 'REMOVE_LIST_ITEM',
+    RESET: 'RESET'
+}
+export const CustomerProductListsContext = createContext(CPLInitialValue)
+// eslint-disable-next-line react/prop-types
+export const CustomerProductListsProvider = ({children}) => {
+    const [state, dispatch] = useReducer((state, {type, payload}) => {
+        switch (type) {
+            case CPLActionTypes.RECEIVE_LISTS: {
+                const productLists = payload.reduce((prev, curr) => {
+                    return {
+                        ...prev,
+                        [curr.id]: curr
+                    }
+                }, {})
+                return {...state, productLists}
+            }
+            case CPLActionTypes.RECEIVE_LIST: {
+                // Tips: if you are unfamiliar with the concept of
+                // reducers, keep in mind that reducers must be pure.
+                // For an action like this, you must update every
+                // level of nested data to avoid unexpected side effects
+                return {
+                    ...state,
+                    productLists: {
+                        ...state.productLists,
+                        [payload.id]: payload
+                    }
+                }
+            }
+            case CPLActionTypes.CREATE_LIST_ITEM: {
+                const items = state.productLists[payload.listId]?.customerProductListItems || []
+
+                // if the item is already added to the list
+                // we update the existing item
+                const existingItemIndex = items.findIndex((item) => item.id === payload.item.id)
+                if (existingItemIndex >= 0) {
+                    items[existingItemIndex] = payload.item
+                } else {
+                    items.push(payload.item)
+                }
+                return {
+                    ...state,
+                    productLists: {
+                        ...state.productLists,
+                        [payload.listId]: {
+                            ...state.productLists[payload.listId],
+                            customerProductListItems: items
+                        }
+                    }
+                }
+            }
+            case CPLActionTypes.UPDATE_LIST_ITEM: {
+                const productLists = {
```

Comment:
NIT: I noticed that you access values from the payload object insitu and choose not to destucture them at the top of the _case_. For example, in this _case_ you access the _listId_ twice and the _item_ twice. I'd recommend that you do do some destructuring at the top of the _case_, this will allow you to do any property validation if we end up going down that route. 

Like this:

```
const {listId, item} = payload

if (!listId) { throw new Error('Missing payload listId') }
if (!item) { throw new Error('Missing payload item') }

```